openssl/src/ssl/connector.rs

```diff
@@ -395,9 +395,10 @@ cfg_if! {
 
             let param = ssl.param_mut();
             param.set_hostflags(X509CheckFlags::NO_PARTIAL_WILDCARDS);
-            match domain.parse() {
-                Ok(ip) => param.set_ip(ip),
-                Err(_) => param.set_host(domain),
+            match (domain.parse(), domain.len()) {
+                (Ok(ip), _) => param.set_ip(ip),
+                (Err(_), 0) => param.set_host(""),
```

Comment:
The problem is in set_host, not here.

Reply:
Agreed :) This is a simple solution, but does not fix the core issue.